### ISSUE TITLE
PSBT: missing witness_utxo for segwit inputs

### ISSUE THREAD

**Issue by bigspider** [spesmilo/electrum#8039]
Currently, electrum seems to be adding the witness_utxo for legacy P2PKH inputs; per BIP-174, the PSBT_IN_WITNESS_UTXO [should only be present for inputs which spend segwit outputs](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#:~:text=This%20should%20only%20be%20present%20for%20inputs%20which%20spend%20segwit%20outputs%2C%20including%20P2SH%20embedded%20ones).

This affects the new Ledger bitcoin app, which relies on this assumption when signing PSBTs with legacy inputs.

**Comment by SomberNight** [comment]
I think we only do this when displaying QR codes. Is that what you mean?
If not, do you have steps to reproduce?

See e.g. this test:
https://github.com/spesmilo/electrum/blob/30ac8896564c0bda22a0c06f2e5973e0270bf6da/electrum/tests/test_wallet_vertical.py#L2654
In the normal flow, where `tx.to_qr_data()` is not called, the input includes `PSBT_IN_NON_WITNESS_UTXO` and lacks `PSBT_IN_WITNESS_UTXO`.

**Comment by bigspider** [comment]
It happened while signing from a standard BIP-44 legacy account; converting the `PartialTransaction` received by the plugin during `sign_transaction` resulted in the following psbt:
```
cHNidP8BAHcCAAAAAXT0yaTajRSLu1boaayjaQ3aDOOsvPgWCcyUbRtvFkrOAAAAAAD9////AhAnAAAAAAAAGXapFCeDycrsILcRWnbAlev4DTq9ohuCiKwkwgAAAAAAABl2qRQ0sv5FvvH8hS5yUi3rDcYH2ol3gIisf0okAAABAPoCAAAAAAEB0OO1kdxITttx6rZqDplHMQB/630IKlDHQvAj6gkBTR4BAAAAFxYAFOMQ0ET4jasbQnaeSoTK8INj+ezB/f///wJg6gAAAAAAABl2qRRFiB7Q01h1UPeUhHt8i5+gPt0KPIisf2YbAAAAAAAXqRTwRk2foOpC2A5NXxRXiDmC4juO7IcCSDBFAiEA4NgNxVQH8N5sYoW6VVzBleEOiUBo35E01ME306Uzi68CIC2B5IlKIAnyqPNVj1wgoj3XHGeNfKzgOBbMAGDITQZoASECqA8AcZTVPTf2+ZU59jU5BYjE4cMosJgpX2GvQNYMsooAAAAAAQEiYOoAAAAAAAAZdqkURYge0NNYdVD3lIR7fIufoD7dCjyIrCIGAtMSuZbG2GtAnMBFWs/KVFyDeyhqZJRU1uLfA7JCoaCaGPWswv0sAACAAQAAgAAAAIAAAAAABgAAAAAiAgN4ZyP/Vn84KuygtxCT15W2TM/5XVIgJ1lm7EurXJmZRxj1rML9LAAAgAEAAIAAAACAAAAAAAcAAAAAIgIDmXMPrZCIByQHALeZ1wN0Rryr8uW9VyY0ruqb/4zVRYcY9azC/SwAAIABAACAAAAAgAEAAAAFAAAAAA==
```

It's from my custom branch, but it shouldn't be related to anything I do, since I create the psbt with `tx.serialize_as_bytes()`.

If it's not enough info to reproduce yet, I should be able to open the initial PR for #7897 tomorrow; it might be easier to reproduce on the PR's branch, since the current Ledger plugin doesn't use psbt.

**Comment by SomberNight** [comment]
Something is weird with your tx. The input has both `PSBT_IN_NON_WITNESS_UTXO` and `PSBT_IN_WITNESS_UTXO` set.
Electrum never does this.
https://github.com/spesmilo/electrum/blob/335f55a7bf07fe9d4634dccf707a659bfd978144/electrum/transaction.py#L1253

**Comment by bigspider** [comment]
Indeed, another weird thing about this psbt is that the both outputs have the BIP32 derivation info, while that's only relevant for the change output.
I made the transaction by sending to a different receiving address from the same wallet (first output); I wonder if that could be the source of the weirdness?

**Comment by SomberNight** [comment]
> both outputs have the BIP32 derivation info, while that's only relevant for the change output.
> I made the transaction by sending to a different receiving address from the same wallet (first output)

That's expected. The wallet adds all the info it has.
I guess it's a design decision and you could argue either way, but I think this behaviour is fine.

No idea about the witness utxo weirdness though.

**Comment by bigspider** [comment]
You are completely right, the `witness_utxo` was added from my plugin.
My mistake, sorry for wasting your time, and thanks for the clarifications!

**Comment by bigspider** [comment]
Reopened, as there might still be an inconsistency with the stamdard. My understanding is the following:
- for legacy inputs, witness_utxo shouldn't be there
- for segwitv0 inputs, **both** non_witness_utxo and witness_utxo should be present. Having the witness_utxo without the non_witness_utxo currently causes a warning with Ledger hardware wallets (and probably other devices, too), because there are some possible attacks.
- for segwitv1 inputs, just the witness_utxo is enough,

Currently, electrum seems to only put the non_witness_utxo and skip the witness_utxo for segwitv0 inputs.

I can easily fix it in the plugin code before sending to the device − but mentioning it anyway as I think it's not matching the expectations of BIP174.

**Comment by SomberNight** [comment]
> for segwitv0 inputs, both non_witness_utxo and witness_utxo should be present. Having the witness_utxo without the non_witness_utxo currently causes a warning with Ledger hardware wallets (and probably other devices, too), because there are some possible attacks.

I don't see why you would want both non_witness_utxo and witness_utxo present at the same time. It is a waste of space and potential for inconsistency (if they mismatch). The witness_utxo can be cleanly calculated from the non_witness_utxo (and the prevout_idx). IIRC Bitcoin Core might be including both but only for compatibility reasons with old versions of itself.

In my reading, [the spec](https://github.com/bitcoin/bips/blob/15c8203eb36304efa1e4588b950f62a5bb32f965/bip-0174.mediawiki#Specification) allows this:

> `PSBT_IN_NON_WITNESS_UTXO`
> This should be present for inputs that spend non-segwit outputs and can be present for inputs that spend segwit outputs. An input **can** have both `PSBT_IN_NON_WITNESS_UTXO` and `PSBT_IN_WITNESS_UTXO`.
>
> `PSBT_IN_WITNESS_UTXO`
> This should only be present for inputs which spend segwit outputs
>
> Why **can** both UTXO types be provided? 
> In order to be compatible with software which were expecting `PSBT_IN_WITNESS_UTXO`

So `PSBT_IN_NON_WITNESS_UTXO` can be present any time, and `PSBT_IN_WITNESS_UTXO` should not be present for legacy inputs. They can both be present but it's redundant.

I mean, if there is a real motivating use case, we could include both, but if it's only old versions of Bitcoin Core then I don't see the point. If the ledger firmware code is expecting both, I would argue it is that firmware that should be changed. :)

**Comment by bigspider** [comment]
The specs is indeed not super explicit. Since PSBT_IN_WITNESS_UTXO can be computed from PSBT_IN_NON_WITNESS_UTXO, it would seem that it's redundant at first sight. My reasoning on why that's not the right reading of the specs is the following:
- the non-witness-utxo is a full transaction, that can be arbitrarily big (imagine if you're spending an output from a transaction out of an exchange that does batching; that could easily be tens of kb, and it all needs to be sent to the signing device!); instead, the witness-utxo is always less 60 bytes.
- segwitv0 tried to fix the problem by making the sighash computation depend on the witness-utxo field, but fell short and there is a bug (arguably not easy to exploit, but a bug nonetheless; see footnote 7 in BIP 174); that's why hardware signers started expecting the non-witness-utxo anyway.
- segwitv1 actually did fix the bug, so it's safe to sign transactions without non-witness-utxo. Therefore, for taproot inputs, the non-witness-utxo _should not_ be included, allowing the psbt to be much smaller.

Because of this, my impression is that relying on the non-witness-utxo should be considered the legacy quirk, and the witness-utxo is the way to go. It's true that the witness utxo is redundant data if the non-witness-utxo is present, but it is very little redundant data − while for segwitv1 the non-witness-utxo includes potentially _a lot_ of useless data.

I'd like to hear the opinion of @achow101 if he wants to chime in.

In the current implementation, the Ledger app always uses the witness utxo for segwitv0 or segwitv1 inputs, but gives a warning for segwitv0 inputs if the non-witness-utxo is missing (to protect from the exploitable edge-cases).
The PR #8041 already includes a workaround (adding the witness-utxo if missing before sending to the device), so it's not really impacting me at this point.
Feel free to close if you deem it not relevant − but I still find the discussion about the specs valuable.

**Comment by achow101** [comment]
Non-witness UTXO is required for non-segwit inputs, and may be included for segwit inputs. Witness UTXO is required for segwit inputs, and should not be included for non-segwit inputs. Segwit inputs can have both fields in order to deal with the sighash issue, but also for backwards compatibility with (versions of) software written before the issue was discovered.

**Comment by SomberNight** [comment]
> Witness UTXO is required for segwit inputs

where does the spec say that? weird that no one complained for years that we are only putting non-witness-utxo for segwit v0 inputs.

**Comment by achow101** [comment]
s/required/expected

The simple signer algorithm described in the BIP looks for the presence of witness UTXO to determine how to sign.

**Comment by SomberNight** [comment]
> The simple signer algorithm described in the BIP looks for the presence of witness UTXO to determine how to sign.

The simple signer algorithm also expects the lack of `non_witness_utxo` for witness v0 inputs. So it is broken regardless and not a good argument for this change.

Anyway, I don't think the spec is explicit about this, but I've found practical examples of incompatibility (https://github.com/cryptoadvance/specter-desktop/issues/868, https://github.com/cryptoadvance/specter-desktop/issues/1046), so I've changed this now in https://github.com/spesmilo/electrum/commit/d3227d7489fe327bd40e891a517c86bd207227ec.
